src/main/java/com/mkopec/ticketbookingapp/service/SeatValidationService.java

```diff
@@ -0,0 +1,199 @@
+        int toIndex = index > lastElementIndex - NUM_OF_NEIGHBOURS ? lastElementIndex : index + 1 + NUM_OF_NEIGHBOURS;
+
+        return models.subList(fromIndex, toIndex);
+    }
+
+    private boolean isReservationValidForSeatModel(List<SeatModel> neighbourhood, SeatModel model) {
+        int modelIndex = neighbourhood.indexOf(model);
+        boolean isFreeSeatLeft = false;
+
+        if (hasPrevious(modelIndex)) {
+            SeatModel p = getPrevious(neighbourhood, modelIndex);
+            switch (p.seatType) {
+                case FREE_INSIDE: {
+                    if (hasPrevious(modelIndex - 1)) {
+                        SeatModel pp = getPrevious(neighbourhood, modelIndex - 1);
+                        switch (pp.seatType) {
+                            case FREE_INSIDE:
+                            case FREE_ON_EDGE:
+                                isFreeSeatLeft = false;
+                                break;
+                            case OCCUPIED:
+                            case RESERVATION_DONE:
+                                isFreeSeatLeft = true;
+                                break;
+                        }
+                    }
+                    break;
+                }
+                case FREE_ON_EDGE:
+                    isFreeSeatLeft = true;
+                    break;
+                case OCCUPIED:
+                case RESERVATION_DONE:
+                    return true;
+            }
+        }
+
+        if (isFreeSeatLeft) {
+            if (hasNext(neighbourhood, modelIndex)) {
+                SeatModel n = getNext(neighbourhood, modelIndex);
+                switch (n.seatType) {
+                    case FREE_INSIDE:
+                    case FREE_ON_EDGE:
+                        return false;
+                    case OCCUPIED:
+                    case RESERVATION_DONE:
+                        break;
+                }
+            }
+        } else {
+            if (hasNext(neighbourhood, modelIndex)) {
+                SeatModel n = getNext(neighbourhood, modelIndex);
+                switch (n.seatType) {
+                    case FREE_INSIDE: {
+                        if (hasNext(neighbourhood, modelIndex + 1)) {
+                            SeatModel nn = getNext(neighbourhood, modelIndex + 1);
+                            switch (nn.seatType) {
+                                case FREE_INSIDE:
+                                case FREE_ON_EDGE:
+                                    break;
+                                case OCCUPIED:
+                                case RESERVATION_DONE:
+                                    return false;
+                            }
+                        }
+                        break;
+                    }
+                    case FREE_ON_EDGE:
+                    case OCCUPIED:
+                    case RESERVATION_DONE:
+                        break;
+                }
+            }
+        }
+
+        return true;
+    }
```

Comment:
ta klasa ma tyle ifów i switchów, że nigdy nie chciałbyś takiej reviewować / utrzymywać.

Reply:
Sporo ifów się zrobiło z powodu iterowania po liście po sąsiednich miejscach (nie zawsze występują w sąsiedztwie dwa na lewo i dwa na prawo jak w ogólnym przypadku). Nie zastosowałem listIterator, ponieważ po wywołaniu next() lub previous() zmienia się index wskazywanego elementu, a tutaj było to niepożądane.